### ISSUE TITLE
polledExpect() can cause Chrome's JS heap to run out of space when running E2E tests

### ISSUE THREAD

**Issue by shermaneric** [blackboard/protractor-sync#31]
**Background**

- When consuming protractor-sync version 5.1.11, we will periodically get failures such as the stack trace below. 

Opening this as an issue for further investigation.
Let me know if you need more informaiton.

Thanks!
Eric

```
00:32:51.750 [chrome #01-51] <--- Last few GCs --->
00:32:51.750 [chrome #01-51] 
00:32:51.750 [chrome #01-51]    70112 ms: Mark-sweep 1304.6 (1434.1) -> 1304.6 (1434.1) MB, 1264.5 / 0.0 ms [allocation failure] [GC in old space requested].
00:32:51.750 [chrome #01-51]    71338 ms: Mark-sweep 1304.6 (1434.1) -> 1304.6 (1434.1) MB, 1225.4 / 0.0 ms [allocation failure] [GC in old space requested].
00:32:51.750 [chrome #01-51]    72641 ms: Mark-sweep 1304.6 (1434.1) -> 1304.6 (1418.1) MB, 1302.7 / 0.0 ms [last resort gc].
00:32:51.750 [chrome #01-51]    73726 ms: Mark-sweep 1304.6 (1418.1) -> 1304.5 (1418.1) MB, 1084.6 / 0.0 ms [last resort gc].
00:32:51.750 [chrome #01-51] 
00:32:51.750 [chrome #01-51] 
00:32:51.750 [chrome #01-51] <--- JS stacktrace --->
00:32:51.750 [chrome #01-51] 
00:32:51.750 [chrome #01-51] ==== JS stack trace =========================================
00:32:51.750 [chrome #01-51] 
00:32:51.750 [chrome #01-51] Security context: 0x26aab64cfb51 <JS Object>
00:32:51.750 [chrome #01-51]     1: compare [/mnt/mesos/sandbox/ultra-ui/node_modules/jasmine/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:2744] [pc=0x23a5fab8bf25] (this=0x2acf46d7c221 <an Object with map 0x2b726c724441>,actual=0,expected=1)
00:32:51.750 [chrome #01-51]     2: toBe [/mnt/mesos/sandbox/ultra-ui/node_modules/protractor-sync/dist/polled-expect.js:~24] [pc=0x23a5fadb98d8] (this=0x14279c2dddc1 <an Object with map 0x2b726c742899>,...
00:32:51.750 [chrome #01-51] 
00:32:51.750 [chrome #01-51] FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
00:32:51.750 [chrome #01-51]  1: node::Abort() [/mnt/mesos/sandbox/ultra-ui/.nvm/versions/node/v6.10.0/bin/node]
00:32:51.750 [chrome #01-51]  2: 0x10983ec [/mnt/mesos/sandbox/ultra-ui/.nvm/versions/node/v6.10.0/bin/node]
00:32:51.750 [chrome #01-51]  3: v8::Utils::ReportApiFailure(char const*, char const*) [/mnt/mesos/sandbox/ultra-ui/.nvm/versions/node/v6.10.0/bin/node]
00:32:51.750 [chrome #01-51]  4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/mnt/mesos/sandbox/ultra-ui/.nvm/versions/node/v6.10.0/bin/node]
00:32:51.750 [chrome #01-51]  5: v8::internal::Handle<v8::internal::AllocationSite> v8::internal::Factory::New<v8::internal::AllocationSite>(v8::internal::Handle<v8::internal::Map>, v8::internal::AllocationSpace) [/mnt/mesos/sandbox/ultra-ui/.nvm/versions/node/v6.10.0/bin/node]
00:32:51.751 [chrome #01-51]  6: v8::internal::Factory::NewAllocationSite() [/mnt/mesos/sandbox/ultra-ui/.nvm/versions/node/v6.10.0/bin/node]
00:32:51.751 [chrome #01-51]  7: v8::internal::AllocationSiteCreationContext::EnterNewScope() [/mnt/mesos/sandbox/ultra-ui/.nvm/versions/node/v6.10.0/bin/node]
00:32:51.751 [chrome #01-51]  8: v8::internal::Runtime_CreateObjectLiteral(int, v8::internal::Object**, v8::internal::Isolate*) [/mnt/mesos/sandbox/ultra-ui/.nvm/versions/node/v6.10.0/bin/node]
00:32:51.751 [chrome #01-51]  9: 0x23a5fa2092a7```

**Comment by scriby** [comment]
Yeah definitely looks like a memory leak. How long do the tests run before hitting this?

**Comment by shermaneric** [comment]
Right.  It hasn't seemed to have happened that often yet.

* Blackboard ultra-ui started consuming open source protractor-sync v5.1.11 yesterday (yay!)
* I only noticed this particular issue in one of the feature branches that had also consumed protractor-sync 5.1.11.  Running the same feature branch again was fine.
* I haven't noticed this particular stack trace pop up again anywhere else.  Caveat being on our /develop branch we noticed [this stack trace](https://github.com/blackboard/protractor-sync/issues/32) once in 10 times since protractor-sync 5.1.11.  I'm assuming those are separate issues for now :)
* Each run above executes roughly 200 E2E tests currently.
* In the failing run with the above stack trace in this issue, this failure happens roughly 32 minutes into the test.  It was the last test group to be executed.  Our average runtimes have been in the 30-35 minute range.

Let me know if you need any other info.  Thanks @scriby!

**Comment by mindywhitsitt** [comment]
I'm not certain that the cause is in polledExpect itself - I suspect that may just be the point where we happened to run out of memory in this particular run.

The protractor-sync upgrade is certainly the most likely culprit since we've never seen this before. I'm curious if you've seen the same thing, @scriby , but I'm assuming you haven't. Ultra has its own memory issues so I imagine we're putting quite a beating on the system when we run our ~200 protractor-sync tests.

**Comment by scriby** [comment]
Yeah, as Mindy said just b/c the failed allocation is in polledExpect doesn't mean that's where the leak is.

By default Node will just use around 1.4 gigs of memory, so the amount of memory Chrome or whatever else on the box is using probably doesn't matter much (assuming the box itself isn't running out of memory).

I'll see if I can repro the memory issue just by running the unit tests a bunch, and if not I'll need to collect some more information.

**Comment by scriby** [comment]
I can see memory leaking when running the tests repeatedly, so at least I can repro the issue.

This looks like it's going to take some time to figure out. In the meantime I recommend increasing node's heap size and/or splitting up the tests across more node processes so each process uses less memory.

**Comment by mindywhitsitt** [comment]
Phew. Glad that you can reproduce it at least. :) I am poking around with it and will let you know if I find anything interesting, but this is not an area I excel in.

**Comment by casteaux** [comment]
i can experiment with increasing the max instances.  We may lose a minute or so on average, but it would be worth it if we don't see failures

**Comment by scriby** [comment]
A little more info... I created a couple test files that just run through a basic test in both vanilla protractor and protractor-sync here: 

https://gist.github.com/scriby/a9da97fc3bfbc3df44037938a036742e

After 5,000 runs, the heap sizes of using vanilla protractor and protractor-sync are very similar:

Vanilla protractor: 84.1 MB -> 97.1 MB
Protractor-sync: 85.4 MB -> 97.4 MB

Both of them exhibit a small leak, but < 10 MB across 5,000 tests is not a big deal.

So it seems most likely that a particular feature or features which aren't part of this simple test are responsible for the leaks. It's also possible that the leaks I see when running the tests are different than what you are seeing at BB.

**Comment by scriby** [comment]
@casteaux how many tests are running per process now?

I am pretty sure the memory leak you are seeing is much more severe than what I see when running the tests. The leak I see is about +3 MB on the heap per run of the unit tests, which is 68 tests. At the same rate that would only account for around 10 MB across 200 tests.

**Comment by casteaux** [comment]
@scriby I bumped the max instances from 4 to 5 and am on my second run without a failure.  I plan to kick it a handful of times over the weekend.  I don't think we can go any higher than 7 without making sacrifices to our runtimes.

**Comment by scriby** [comment]
Even at 50 tests per instance, 1.4 GB of memory from node is pretty absurd. There is something pretty wrong going on there.

I would recommend taking some memory dumps at some interval while the tests are running (you can see a technique for doing so in the gist I posted above). You can open the memory dumps in the Chrome dev tools memory pane. Load multiple dumps, and then you can use the "Comparison" option in the first dropdown to get a diff between them, which helps get an idea of what's being retained.

Also https://www.dynatrace.com/blog/understanding-garbage-collection-and-hunting-memory-leaks-in-node-js/ is good for more details on that.

**Comment by casteaux** [comment]
Thanks, @scriby!  I'll check that out

**Comment by mindywhitsitt** [comment]
I copied your code, @scriby and took a bunch of heap snapshots while the tests were running. I'm wondering if I did something wrong though, because the size doesn't really appear to be growing.
<img width="256" alt="screen shot 2017-07-17 at 4 02 27 pm" src="https://user-images.githubusercontent.com/2130396/28287378-75146da6-6b09-11e7-9113-71a86fec2834.png">

**Comment by casteaux** [comment]
Now we're talkin'

<img width="683" alt="screen shot 2017-07-17 at 5 01 01 pm" src="https://user-images.githubusercontent.com/6486455/28289688-900f8cf0-6b11-11e7-9325-aef46216edd0.png">

**Comment by casteaux** [comment]
continued ...
<img width="697" alt="screen shot 2017-07-17 at 5 26 48 pm" src="https://user-images.githubusercontent.com/6486455/28290625-279c0514-6b15-11e7-970f-59d55fe21521.png">

@scriby I updated your code to use protractor-sync's `polledExpect()` in place of `expect()`, but I think that's the only real difference

**Comment by mindywhitsitt** [comment]
I see that polledExpect has some setTimeout calls. Are they getting cleared?

**Comment by scriby** [comment]
Good catch! I'll do some test runs with polledExpect as well.

**Comment by scriby** [comment]
Hmm I changed the expects to polledExpects in the protractor-sync_test.ts and the results are pretty much the same as before.

I'm seeing heap usage go from 65.9 MB -> 81.2 MB over 5,000 runs.

I'm using node v6.11.0 on OSX.

@mindywhitsitt the setTimeout in polledExpect doesn't specify a wait time, so it executes during the next event loop.

**Comment by mindywhitsitt** [comment]
Ah. I was confusing setTimeout with setInterval. 😛

**Comment by mindywhitsitt** [comment]
I confirmed that I can get Scriby's test script to grow the heap for me also, even if I use it as-is (without replacing polledExpect). And since it's really not growing for you, then maybe we need to look more closely at our runner/framework code at BB.

**Comment by scriby** [comment]
@mindywhitsitt @casteaux the way I was testing this was by adding this test file into the protractor-sync repo's test folder and running it directly from here. Are you guys testing the same way or testing it from within the BB testing framework?

Just to compare notes, I'm running with protractor 5.1.2 and grunt-protractor-runner 5.0.0 as specified in the package.json of this repo.

**Comment by mindywhitsitt** [comment]
When I tested earlier, I tried dropping the files in with Ultra's tests and running them from there. This morning I did what you described: dropped them in with protractor-sync's other tests under test/spec. I had to import profiler as `var profiler = require('v8-profiler');` because I got a type error when I used the other import style, and I couldn't figure out how to expose gc properly in grunt-protractor-runner so I hacked into it and added `--expose-gc` when it runs protractor.

Same grunt-protractor-runner and protractor versions as you. My node version is v6.10.1

My heap snapshots from this morning show memory increasing from 145MB at the first snapshot to 178MB at the last snapshot with both protractor-sync and protractor. That's so far off from your numbers that I'm seriously wondering if I did it right. :/

**Comment by casteaux** [comment]
That's really close to what I did as well: dropped my test under test/spec, imported v8-profiler, and stole mindy's gc hack.  My heap snapshots went from 84MB to 241MB in < 1 hour.

**Comment by casteaux** [comment]
@scriby i ran this script in the protractor-sync project and saw similar behavior
<img width="770" alt="screen shot 2017-07-18 at 2 35 32 pm" src="https://user-images.githubusercontent.com/6486455/28333635-7c3493aa-6bc6-11e7-8586-f96572cf1ce0.png">

**Comment by mindywhitsitt** [comment]
@scriby I tried sending 2 heap snapshots to you. @casteaux took them while running your sample tests within protractor sync.

I'm pretty confused by the fact that you aren't seeing the same errors as we are, and I'm tempted to conclude that we're imagining things, except the build server really is running out of memory every 10-20 runs or so. :P

**Comment by mindywhitsitt** [comment]
Still working on tracking down this leak. I'll leave the issue open for a while longer in the hopes that we can pin it down and clear Protractor-sync's name. :)

@scriby , I am copying your code to take heap snapshots, and I was wondering if you would explain to me why you wrapped part of it in setImmediate:

````
setImmediate(() => {
  global.gc();
  const snap = profiler.takeSnapshot('profile');
  saveHeapSnapshot(snap, '/Users/scriby/heaps');
});
````

Will something bad happen if I don't use that?

**Comment by scriby** [comment]
The setImmediate was in the article I linked above. I've never used it before, but I can see the logic in using it to have the memory dump run after the current pass through the event loop finishes. Also, I couldn't see any difference calling gc.collect manually (it's possible the act of taking a snapshot does a collection), so you may not need to go to great lengths to enable that.

@casteaux you mentioned that you see the heap grow to 241 MB after around an hour. When I did my tests, the 5,000 iterations finished after about 7 minutes. If I increased the loop count and let it run I could probably see a similar heap growth over the same time frame.

However, I'm not convinced "that" heap growth is the same you are seeing when running the real tests for the BB E2E tests. You wouldn't even notice the 241 MB overhead. Something is leaking 1.4 GB of memory within about 50 test runs, which is rather remarkable.

@mindywhitsitt @casteaux are either of you able to share heap snapshots taken while running the BB tests when they are taking up several hundred MB of memory? I can take a quick peek and see if anything jumps out.

Other things you could start thinking about would be to run individual files and watch memory growth to start "binary searching" which tests cause the leak. If all tests seem to cause it, then that also gives us information and from there you could run a single test repeatedly and start removing things from that test until the leak goes away.

**Comment by mindywhitsitt** [comment]
Thanks for this!

Yes, it took a little while to realize that the heap growth we've been able to capture so far is likely not the same as whatever is bringing the tests down. Now that we know we are hunting something else, I have been trying for the past couple of days to capture a heap snapshot at the time that the memory usage starts to grow out of control, and so far I haven't gotten anything over 200MB, so I believe I haven't found it yet.

I have a few continuous runs set up on the build server trying to catch the problem. What I really want is a way to take a heap snapshot right before Node crashes with the out of memory exception, but it looks like Node doesn't provide that. I'm using node-gc-profiler (https://github.com/bretcope/node-gc-profiler) right now to check memory usage each time the garbage collector runs and take a snapshot if it's over a certain threshold.

I think I'll also take your advice and set up some of those binary searches on the build server. That sounds like a really good idea.

**Comment by mindywhitsitt** [comment]
Fixed in PR #37
